File: modules/wem-api/src/main/java/com/enonic/wem/api/content/page/layout/LayoutRegions.java
package com.enonic.wem.api.content.page.layout;


import com.enonic.wem.api.content.page.AbstractRegions;
import com.enonic.wem.api.content.page.ComponentPath;
import com.enonic.wem.api.content.page.region.Region;

public final class LayoutRegions
    extends AbstractRegions
{
    private LayoutRegions( final Builder builder )
    {
        super( builder );
    }

    public void applyComponentPaths( final ComponentPath parent )
    {
        for ( final Region region : this )
        {
            region.applyComponentPaths( parent );
        }
    }

    public static Builder newLayoutRegions()
    {
        return new Builder();
    }

    public static class Builder
        extends AbstractRegions.Builder<Builder>
    {
        public LayoutRegions build()
        {
            return new LayoutRegions( this );
        }
    }
}

File: modules/wem-webapp/src/main/webapp/admin/common/js/content/page/layout/LayoutRegions.ts
module api.content.page.layout {

    export class LayoutRegions extends api.content.page.AbstractRegions {

        constructor(builder: LayoutRegionsBuilder) {

            super(builder.regions);
        }
    }

    export class LayoutRegionsBuilder {

        regions: api.content.page.region.Region[] = [];

        fromJson(regionsJson: api.content.page.region.RegionJson[], layoutComponent:ComponentPath): LayoutRegionsBuilder {

            regionsJson.forEach((regionJson: api.content.page.region.RegionJson) => {

                var regionPath = new RegionPath(layoutComponent, regionJson.name);

                var regionBuilder = new api.content.page.region.RegionBuilder().
                    setName(regionJson.name).
                    setPath(regionPath);

                regionJson.components.forEach((componentJson: api.content.page.PageComponentTypeWrapperJson) => {
                    var pageComponent = api.content.page.PageComponentFactory.createFromJson(componentJson, regionPath);
                    regionBuilder.addComponent(pageComponent);
                });


                this.addRegion(regionBuilder.build());
            });
            return this;
        }

        addRegion(value: api.content.page.region.Region): LayoutRegionsBuilder {
            this.regions.push(value);
            return this;
        }

        public build(): LayoutRegions {
            return new LayoutRegions(this);
        }
    }
}